### ISSUE TITLE
Dropped subscriptions still receive events

### ISSUE THREAD

**Issue by urbanhusky** [EventStore/EventStore#929]
I've set up an all subscription on a connection with a very low timeout for testing purposes. 

**Steps to reproduce**
1. Connect with heartbeat and operation timeout of 5 seconds
2. `SubscribeToAllFrom` using `Position.Start`. Recreate subscription when dropped
3. Pause debugging
4. Resume debugging

**Effects**

Subscriptions are dropped and `subscriptionDropped` is invoked, thus recreating the subscription.
Any further events now invoke both subscriptions in parallel.

**Example code on how I create a subscription**

``` C#
private void CreateSubscription(int index, string subscriptionName)
{
    Output("Creating subscription for #" + index + " named " + subscriptionName);
    var tracker = new PerformanceTracker("#" + index + "_" + subscriptionName);

    this.connection.SubscribeToAllFrom(
        Position.Start,
        new CatchUpSubscriptionSettings(1000, 1000, true, true),
        (s, evt) =>
        {
            if (evt.Event.EventType.StartsWith("$"))
            {
                return; // skip system events
            }

            Debug.WriteLine($"{index:N0}{subscriptionName} got event " + evt.Event.EventType + " (" + evt.Event.EventId.ToString() + ")");
            tracker.Increment();
        },
        subscriptionDropped: (subscription, reason, exception) =>
        {
            tracker.Resubscribed();
            CreateSubscription(index, "Re:" + subscriptionName);
            GenerateEvents(1, "subscriptionDropped_" + index);
        },
        liveProcessingStarted: (s) =>
        {
            Output(tracker.Name + " started live processing");
            GenerateEvents(1, "liveProcessingStarted" + index);
        },
        userCredentials: new UserCredentials("admin", "changeit"));
}
```

With `index` being just a number used to test multiple parallel subscriptions, not the position in the stream.

**Example output**

```
Creating subscription for #0 named Initial
0Initial got event demoEvent.Init (defd0ec8-37bb-4ce8-92ba-d05ed6ec8cc4)
0Initial got event demoEvent.Init (f64cd55f-19d8-4a25-9db3-159bba4ef063)
0Initial got event demoEvent.Init (0d3d4e07-0631-4fd5-8b0c-16e298337be2)
0Initial got event demoEvent.Init (7cd05981-d072-4907-aa9c-33d20730bc05)
0Initial got event demoEvent.Init (ffc4f61f-e1eb-4647-b55b-3805f6309e5f)
#0_Initial started live processing
0Initial got event demoEvent.liveProcessingStarted0 (4b18f329-15bb-4372-a814-73371d8fabeb)
```

_...pause debugging..._

```
#0_Initial resubscribed (should drop to new name)
Creating subscription for #0 named Re:Initial
0Re:Initial got event demoEvent.Init (defd0ec8-37bb-4ce8-92ba-d05ed6ec8cc4)
0Re:Initial got event demoEvent.Init (f64cd55f-19d8-4a25-9db3-159bba4ef063)
0Re:Initial got event demoEvent.Init (0d3d4e07-0631-4fd5-8b0c-16e298337be2)
0Re:Initial got event demoEvent.Init (7cd05981-d072-4907-aa9c-33d20730bc05)
0Re:Initial got event demoEvent.Init (ffc4f61f-e1eb-4647-b55b-3805f6309e5f)
0Re:Initial got event demoEvent.liveProcessingStarted0 (4b18f329-15bb-4372-a814-73371d8fabeb)
0Initial got event demoEvent.subscriptionDropped_0 (8cc947b5-dca0-4cbb-9ca1-84cac6ab8764)
#0_Initial --- incremented even though resubscribed! ---
#0_Initial started live processing
0Re:Initial got event demoEvent.subscriptionDropped_0 (8cc947b5-dca0-4cbb-9ca1-84cac6ab8764)
#0_Re:Initial started live processing
0Re:Initial got event demoEvent.liveProcessingStarted0 (91001bee-83c0-4e80-8cd4-934ceafbe1e6)
0Initial got event demoEvent.liveProcessingStarted0 (91001bee-83c0-4e80-8cd4-934ceafbe1e6)
0Re:Initial got event demoEvent.liveProcessingStarted0 (29ea3bce-7c58-4e5e-bb35-b4fb02b46e57)
#0_Initial --- incremented even though resubscribed! ---
0Initial got event demoEvent.liveProcessingStarted0 (29ea3bce-7c58-4e5e-bb35-b4fb02b46e57)
#0_Initial --- incremented even though resubscribed! ---
```

The interesting thing is also that `subscriptionDropped` is called at most once per subscription. `liveProcessingStarted` however is called multiple times.

P.S.: Yes, I'm not tracking the position so the recreated subscription reads all events. I merely didn't implement that in this demo code :)

**Comment by hayley-jean** [comment]
Apologies for taking so long to reply

This appears to be the same issue as the one discussed on google groups [here](https://groups.google.com/forum/#!searchin/event-store/subscription/event-store/AdKzv8TxabM/6RzudeuAAgAJ), which we still need to look into.

As a workaround for now, you can simply call `Stop()` on the subscription in the `SubscriptionDropped` callback to prevent the dropped subscriptions from getting new events (see Greg's code sample in the google groups discussion)

**Comment by hayley-jean** [comment]
Fixing this properly will break client functionality for subscriptions. As such, we will only be able to fix it in version 4.0.0.

**Comment by Narvalex** [comment]
Was this issue fixed in version 4.0.0? This issue is critical beacause I get leaked handlers or no handlers at all!

**Comment by Narvalex** [comment]
Maybe is because I do not know the exactly use of KeepReconnecting() and KeepRetrying()

**Comment by Narvalex** [comment]
Also, a confusing stuff is how to handle the subscription.Stop(timeout parameter) when timeout is reached.

**Comment by hayley-jean** [comment]
Hi @Narvalex 

This hasn't been fixed yet. For now, call `Stop()` in your connection dropped handler and then resubscribe.

With regards to `subscription.Stop(Timeout)`, when the timeout is reached, it will throw a timeout exception.

**Comment by Narvalex** [comment]
Thanks @hayley-jean! Just to be sure... Is it guaranteed that no new event will be passed to the handler when the dropped handler is called?

**Comment by hayley-jean** [comment]
If you have called `subscription.Stop()` then no new events will be handled by that subscription.

My apologies, but I've just noticed that above I said you should call `Stop()` in your `connection dropped` handler. That should actually be in your `subscription dropped` handler (not the connection). I've edited it now.

**Comment by Narvalex** [comment]
Thanks @hayley-jean! My concern is that I do not want to leak handlers. Since there is a chance that when a non-thread-safe handler is handling an event at the same time a disconnection ocurrs, then I am calling Stop() without timeout checking and then creating a new subscription... that will be leaking a new handler if the old handler is still working... I added messy pesimistic locks/semaphores to avoid that...Now I am worry about those things running in producction.

**Comment by MetaFight** [comment]
What's the status of this issue?  I'm using catch-up subscriptions and, even after their Dropped handler has been called they continue to receive events.

This is a problem because, well, when the Dropped handler is called I'm assuming the subscription to be dead, so a create a new one to resume the work.  And then I find myself with two subscriptions doubling up.

I've read that calling `.Stop()` on the subscription is the solution, however this just deadlocks for me and using the overload with a timeout _always_ times out.

Any suggestions?

**Comment by condron** [comment]
Hi @MetaFight just looked through the code and what is happening here is the subscription is auto reconnecting. 

So when you create a new subscription and then the old subscription(s) reconnect you get multiple subscriptions. If you want to make this explicit in the code you can set `.KeepReconnecting()` as an option in `ConnectionSettings`.

I'm not sure why `.Stop()` is hanging, when / how are you calling it?

Overall though this is issue is by design in the code, and needs clearer documentation to explain the behavior, and subsequent drop notifications after the first reconnect do not get sent for some reason.

**Comment by MetaFight** [comment]
Hi @condron.  Thanks for the explanation.

Now that I know this is by design I can (more confidently) let the ES library do its thing and handle sub reconnections for me.

Regarding the `.Stop()` methods, I was calling them from the SubscriptionDropped handler.  I vaguely recall a Google Groups thread where Greg Young said this wouldn't work because `.Stop()` waits for that handler to complete (hence the deadlock/timeouts).  I *thought* a fix had be deployed to make this possible, but I guess not.

Thanks for you help!

**Comment by condron** [comment]
In my test against the latest code the `.Stop()` call in the dropped handler worked fine, what version are you working with?

**Comment by MetaFight** [comment]
Sorry, just noticed your reply.

I'm using version 5.0.1 of the [nuget pagkage](https://www.nuget.org/packages/EventStore.Client/5.0.1/).
My application is a .NET Core 2.0 ASP.Net Core service.
My EventStore instance is version 4.1.0.0.

My code:

```
this.connection = EventStoreConnection.Create(new Uri($"tcp://{eventStoreHost}:1113"));
var subscription =
                this.connection.SubscribeToStreamFrom(
                    stream,
                    startPosition,
                    CatchUpSubscriptionSettings.Default,
                    this.ResolveAndInvokeHandlers,
                    null,
                    this.OnSubscriptionDropped);
// ...

private void OnSubscriptionDropped(EventStoreCatchUpSubscription subscription, SubscriptionDropReason reason, Exception exception)
{
    this.logger.LogWarning($"Subscription to {subscription.StreamId} dropped.  Reason: {Enum.GetName(typeof(SubscriptionDropReason), reason)}.  Exception: {exception.ToString()}");

    this.logger.LogDebug($"Killing subscription...");
    subscription.Stop(TimeSpan.FromMinutes(1));
    this.logger.LogDebug($"...subscription killed.");

    // Recreate subscription.
    this.CreateCatchUpSubscription(subscription.StreamId);
}

```

**Comment by MetaFight** [comment]
I sidestepped the issue with the following change.
```C#
private async void OnSubscriptionDropped(EventStoreCatchUpSubscription subscription, SubscriptionDropReason reason, Exception exception)
{
    this.logger.LogWarning($"Subscription to {subscription.StreamId} dropped.  Reason: {Enum.GetName(typeof(SubscriptionDropReason), reason)}.  Exception: {exception.ToString()}");

    await Task.Yield();
    subscription.Stop(TimeSpan.FromMinutes(1));

    // Recreate subscription.
    this.CreateCatchUpSubscription(subscription.StreamId);
}
```

I'll report back if I have issues with this.  Cheers.

**Comment by p0onage** [comment]
@MetaFight after calling Stop I've still noticed TCP connections left open from the web app and they keep building up after every connect?

**Comment by roblohmann** [comment]
I'm also running into this issue while using persistentsubscriptions. For now we appear to be able to work around this by explicitly closing the connection to the eventstore ourselves and reconnecting.

I haven't fully tested it, but the first results look good. So I'm not aware yet if there are any sideeffect. If so, I'll let you know.

This is done with the following code;

```
//--SNIP
_persistentSubscription = await _connection.ConnectToPersistentSubscriptionAsync(Constants.Stream
                                , Constants.Group
                                , EventAppeared, SubscriptionDropped, null, bufferSize, autoAcknowledge);

//--SNIP
```
and this is the implementation for our SubscriptionDropped method
```
private void SubscriptionDropped(EventStorePersistentSubscriptionBase arg1, SubscriptionDropReason arg2, Exception arg3)
        {
            _logger.LogInformation($"Persistent subscription dropped. reason: {arg2}");
            Console.WriteLine($"Persistent subscription dropped. reason: {arg2}");

            if(arg2 == SubscriptionDropReason.ConnectionClosed)
            {
                _connection.Close();
                _connection.Dispose();
                _connection = null;

                _persistentSubscription = null;
                ConnectAndSubscribe().ConfigureAwait(false).GetAwaiter().GetResult();

                return;
            }

            _persistentSubscription.Stop(TimeSpan.FromMinutes(1));
            _persistentSubscription = null;

            _logger.LogInformation("Actor reconnecting to eventstore");
            Console.WriteLine("Actor reconnecting to eventstore");

            ConnectAndSubscribe().ConfigureAwait(false).GetAwaiter().GetResult();
        }
```

**Comment by jageall** [comment]
This should be fixed in the 5.0.8 client. Note that subscriptions do not automatically reconnect anymore and you will need to resubscribe manually.
